File: Aether.MacOS/Aether/ViewModels/SearchViewModel.swift
import AetherIPC
import Combine
import Foundation
import GRPCCore
import NIO

@MainActor
class SearchViewModel: ObservableObject {
    @Published var query: String = ""
    @Published var results: [GameViewModel] = []
    @Published var totalMatches: Int = 0
    @Published var isSearching: Bool = false
    @Published var errorMessage: String?

    // Filters
    // @Published var filterInstalled: Bool = false // Removed
    @Published var filterPlatform: String? = nil  // nil = all
    @Published var filterGenre: String? = nil

    // Sorting
    @Published var sortBy: Aether_LibrarySearchRequest.SortOption = .relevance
    @Published var sortAscending: Bool = false

    // Reference to AppState for plugins
    var appState: AppState?

    var availableImporters: [PluginViewModel] {
        appState?.plugins.filter { $0.isImporter } ?? []
    }

    private var cancellables = Set<AnyCancellable>()
    private var searchDebounceTask: Task<Void, Never>?

    init() {
        // Debounce search query
        $query
            .removeDuplicates()
            .debounce(for: .milliseconds(300), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                self?.performSearch()
            }
            .store(in: &cancellables)

        // Trigger search on filter changes
        Publishers.CombineLatest3($filterPlatform, $filterGenre, $sortBy)
            .sink { [weak self] _ in
                self?.performSearch()
            }
            .store(in: &cancellables)

        $sortAscending
            .sink { [weak self] _ in
                self?.performSearch()
            }
            .store(in: &cancellables)
    }

    func performSearch() {
        searchDebounceTask?.cancel()

        // Don't search if empty query unless a filter is active
        if query.isEmpty && filterPlatform == nil && filterGenre == nil {
            self.results = []
            return
        }

        searchDebounceTask = Task {
            do {
                self.isSearching = true
                self.errorMessage = nil

                var request = Aether_LibrarySearchRequest()
                request.query = query
                // request.filterInstalled = filterInstalled // Removed
                if let platform = filterPlatform {
                    request.filterPlatforms = [platform]
                }
                if let genre = filterGenre {
                    request.filterGenres = [genre]
                }
                request.sortBy = sortBy
                request.sortAscending = sortAscending
                request.limit = 50

                let response = try await GrpcClient.shared.client.searchLibrary(request)

                // Check cancellation
                if Task.isCancelled { return }

                self.results = response.games.map { GameViewModel(from: $0) }
                self.totalMatches = Int(response.totalMatches)
                self.isSearching = false

            } catch {
                if !Task.isCancelled {
                    self.isSearching = false
                    self.errorMessage = error.localizedDescription
                    print("Search Error: \(error)")
                }
            }
        }
    }

    func clearFilters() {
        filterPlatform = nil
        filterGenre = nil
        query = ""
    }
}

File: Aether.WinUI/Aether.WinUI/ViewModels/SearchViewModel.cs
using Aether.Protos;
using Aether.WinUI.Models;
using Aether.WinUI.Services;
using CommunityToolkit.Mvvm.ComponentModel;
using Microsoft.UI.Dispatching;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Aether.WinUI.ViewModels;

public partial class SearchViewModel : ObservableObject
{
    private readonly GrpcClientService _grpc;
    private readonly MainViewModel _mainViewModel;
    private readonly DispatcherQueue _dispatcher;
    private readonly ILogger<SearchViewModel> _logger;

    private CancellationTokenSource? _searchCts;
    private NotifyCollectionChangedEventHandler? _pluginsChangedHandler;

    [ObservableProperty] private string query = "";
    [ObservableProperty] private ObservableCollection<GameViewModel> results = new();
    [ObservableProperty] private int totalMatches;
    [ObservableProperty] private bool isSearching;
    [ObservableProperty] private string? errorMessage;

    // Filters
    [ObservableProperty] private string? filterPlatform;
    [ObservableProperty] private string? filterGenre;

    // Sorting
    [ObservableProperty] private LibrarySearchRequest.Types.SortOption sortBy = LibrarySearchRequest.Types.SortOption.Relevance;
    [ObservableProperty] private bool sortAscending;

    [ObservableProperty] private bool showEmptyState;
    [ObservableProperty] private bool showProgress;

    public ObservableCollection<PluginViewModel> AvailableImporters { get; } = new();

    public SearchViewModel(GrpcClientService grpc, MainViewModel mainViewModel, ILogger<SearchViewModel> logger)
    {
        _grpc = grpc;
        _mainViewModel = mainViewModel;
        _logger = logger;
        _dispatcher = App.Current?.MainWindow?.DispatcherQueue ?? DispatcherQueue.GetForCurrentThread();

        _logger.LogDebug("SearchViewModel initialized");
        HookPluginCollection();
        RefreshAvailableImporters();
    }

    public int SortByIndex
    {
        get => SortBy switch
        {
            LibrarySearchRequest.Types.SortOption.Name => 1,
            LibrarySearchRequest.Types.SortOption.ReleaseDate => 2,
            LibrarySearchRequest.Types.SortOption.Playtime => 3,
            _ => 0
        };
        set
        {
            SortBy = value switch
            {
                1 => LibrarySearchRequest.Types.SortOption.Name,
                2 => LibrarySearchRequest.Types.SortOption.ReleaseDate,
                3 => LibrarySearchRequest.Types.SortOption.Playtime,
                _ => LibrarySearchRequest.Types.SortOption.Relevance
            };
            OnPropertyChanged(nameof(SortByIndex));
        }
    }

    partial void OnQueryChanged(string value)
    {
        _logger.LogTrace("Query changed: {Query}", value);
        DebounceSearch();
        UpdateEmptyState();
    }

    partial void OnFilterPlatformChanged(string? value)
    {
        _logger.LogTrace("FilterPlatform changed: {Filter}", value);
        DebounceSearch();
    }

    partial void OnFilterGenreChanged(string? value)
    {
        _logger.LogTrace("FilterGenre changed: {Filter}", value);
        DebounceSearch();
    }

    partial void OnSortByChanged(LibrarySearchRequest.Types.SortOption value)
    {
        _logger.LogTrace("SortBy changed: {SortBy}", value);
        DebounceSearch();
        OnPropertyChanged(nameof(SortByIndex));
    }

    partial void OnSortAscendingChanged(bool value)
    {
        _logger.LogTrace("SortAscending changed: {Value}", value);
        DebounceSearch();
    }

    public void ClearFilters()
    {
        _logger.LogDebug("ClearFilters invoked");
        FilterPlatform = null;
        FilterGenre = null;
    }

    public void ClearSearch()
    {
        _logger.LogDebug("ClearSearch invoked");
        _searchCts?.Cancel();
        Query = string.Empty;
        FilterPlatform = null;
        FilterGenre = null;
        Results.Clear();
        TotalMatches = 0;
        ErrorMessage = null;
        IsSearching = false;
        ShowProgress = false;
        ShowEmptyState = false;
    }

    public void TogglePlatformFilter(string platform)
    {
        _logger.LogDebug("TogglePlatformFilter invoked: {Platform}", platform);
        if (string.Equals(FilterPlatform, platform, StringComparison.OrdinalIgnoreCase))
        {
            FilterPlatform = null;
        }
        else
        {
            FilterPlatform = platform;
        }
    }

    private void DebounceSearch()
    {
        _logger.LogTrace("DebounceSearch invoked");
        _searchCts?.Cancel();

        if (string.IsNullOrWhiteSpace(Query) && FilterPlatform == null && FilterGenre == null)
        {
            Results.Clear();
            TotalMatches = 0;
            UpdateEmptyState();
            return;
        }

        _searchCts = new CancellationTokenSource();
        var token = _searchCts.Token;

        _ = Task.Run(async () =>
        {
            try
            {
                await Task.Delay(300, token);
                if (token.IsCancellationRequested) return;

                await PerformSearchAsync(token);
            }
            catch (TaskCanceledException)
            {
                // Ignore
            }
        }, token);
    }

    private async Task PerformSearchAsync(CancellationToken token)
    {
        _logger.LogInformation("PerformSearchAsync invoked");
        try
        {
            await EnqueueAsync(() =>
            {
                IsSearching = true;
                ShowProgress = Results.Count == 0;
                ErrorMessage = null;
            });

            var request = new LibrarySearchRequest
            {
                Query = Query ?? string.Empty,
                SortBy = SortBy,
                SortAscending = SortAscending,
                Limit = 50
            };

            if (!string.IsNullOrWhiteSpace(FilterPlatform))
            {
                request.FilterPlatforms.Add(FilterPlatform);
            }

            if (!string.IsNullOrWhiteSpace(FilterGenre))
            {
                request.FilterGenres.Add(FilterGenre);
            }

            var response = await _grpc.Client.SearchLibraryAsync(request, cancellationToken: token);

            if (token.IsCancellationRequested) return;

            var mapped = response.Games.Select(GameViewModel.FromProto).ToList();

            await EnqueueAsync(() =>
            {
                Results = new ObservableCollection<GameViewModel>(mapped);
                TotalMatches = response.TotalMatches;
                IsSearching = false;
                ShowProgress = false;
                UpdateEmptyState();
            });
        }
        catch (Exception ex)
        {
            if (token.IsCancellationRequested) return;

            await EnqueueAsync(() =>
            {
                IsSearching = false;
                ShowProgress = false;
                ErrorMessage = ex.Message;
                UpdateEmptyState();
            });
            _logger.LogError(ex, "Search failed");
        }
    }

    private void HookPluginCollection()
    {
        _logger.LogDebug("HookPluginCollection invoked");
        _mainViewModel.PropertyChanged += (_, args) =>
        {
            if (args.PropertyName == nameof(MainViewModel.Plugins))
            {
                RefreshAvailableImporters();
                AttachPluginsCollection();
            }
        };

        AttachPluginsCollection();
    }

    private void AttachPluginsCollection()
    {
        _logger.LogTrace("AttachPluginsCollection invoked");
        if (_pluginsChangedHandler != null)
        {
            _mainViewModel.Plugins.CollectionChanged -= _pluginsChangedHandler;
        }

        _pluginsChangedHandler = (_, _) => RefreshAvailableImporters();
        _mainViewModel.Plugins.CollectionChanged += _pluginsChangedHandler;
    }

    private void RefreshAvailableImporters()
    {
        _logger.LogDebug("RefreshAvailableImporters invoked");
        var importers = _mainViewModel.Plugins
            .Where(p => p.Capabilities.Contains("Importer"))
            .OrderBy(p => p.Name)
            .ToList();

        _ = EnqueueAsync(() =>
        {
            AvailableImporters.Clear();
            foreach (var plugin in importers)
            {
                AvailableImporters.Add(plugin);
            }
        });
    }

    private void UpdateEmptyState()
    {
        _logger.LogTrace("UpdateEmptyState invoked");
        ShowEmptyState = !IsSearching && Results.Count == 0 && !string.IsNullOrWhiteSpace(Query);
    }

    private Task EnqueueAsync(Action action)
    {
        _logger.LogTrace("EnqueueAsync invoked");
        var tcs = new TaskCompletionSource<bool>();
        _dispatcher.TryEnqueue(() =>
        {
            action();
            tcs.SetResult(true);
        });
        return tcs.Task;
    }
}
